### ISSUE TITLE
New cmake Supporting "in distro" building mode as well as "à la" 5.2 modes

### ISSUE THREAD

**Issue by dominig** [seandepagnier/climatology_pi#73]
Changes in cmake file done in June by Rick Gleason (which are indicated to originate from Jon Gough) were breaking the "in distro" building of climatology.
In distro does not build in a git env and does not post the result in a compatible way.
As this new cmakefiles have been reused for many plugins, the issue has spread to many other plugins.

I have created an updated version that is posted on my branch https://github.com/dominig/climatology_pi.git that allows to support both modes. The change is control by a new cmake option BUILD_LEGACY_MODE.
@rgleason 
@jongough 
@seandepagnier 
I would be interested to get your feedback as well as confirmation that I have not broken anything before proposing similar changes to other broken plugins.
It could valuable to go a step further and isolate more the different packaging mode in independent cmake files.
Thoughts
Dominig

**Comment by rgleason** [comment]
Jon and Doming, I don't really understand the problem.  We do provide legacy packages in the Cloudsmith repository
https://cloudsmith.io/~opencpn/repos/celestial-navigation-alpha/packages/

Note the files beginning with "opencpn-xxxxx"    for 2.3.3.14 version. There are windows exe, many versions for debian deb and  macos.pkg which all fit within the "Legacy plugin installation" routine.   So I do not understand the problem.

I notice that there are a lot of changes compared to my fork and I am not sure I want to incorporate these right now.

Additionally, we need to update climatology frontend to the current version found in jon's Testplugin, because there have been a number of necessary changes.

@jongough  @dominig 
Let's discuss this further to understand what needs to be done.  I could make the changes and then Doming could rebase his? Thanks

**Comment by rgleason** [comment]
@jongough   
@seandepagnier 
Addiitionally  I have these two problems with climatology
https://github.com/OpenCPN/plugins/issues/264
https://github.com/OpenCPN/plugins/issues/255

**Comment by jongough** [comment]
I am not sure what is intended by this change. It just seems to remove PluginPackage.cmake. The current frontend2 (1.0.143.2) should enable you to build locally, package locally, build in the cloud without deploy, build selected objects in the cloud, deploy both plugin manager and legacy plugin installables, use OCPN or your own cloudsmith repositories, etc. The construct of the cmake files is designed to allow all plugin changes to be constrained to the CMakeLists.txt file and the inclusion of plugin specific cmake files, frontend2 changes are only made to the files marked not to be changed so that they can be directly replaced.

Is there something else that should be done?

**Comment by rgleason** [comment]
That's what I think, but I need to update the frontend. 
It should be clear that we have not tested all environments, that is impossible. We expect users to do this and appreciate these reports.  After I update, please try the plugiin and advise.

**Comment by dominig** [comment]
The issue with the existing cmake is that it requires to build in a git tree.
When you build "in distro" with OBS for OpenSUSE or with Koji for Fedora/Centos the building process is not done in a git tree and the packaging and signing of the rpm is done by the distro build system.
So I have deactivated the requirement to run in a git tree and the bit of cmake which expect info from git when building in legacy mode. 
I am open on the organisation of the cmake to miminise long term maintenance but I need to be able to build without any git requirement of connection to external web site for getting source code of publishing results.

**Comment by jongough** [comment]
All the plugins and OCPN are git based environments. If you are trying to use any of this code without git behind the scenes then you are likely going to have issues with being out of date with the current branches. The frontend2 cmake process only uses the local git to workout what is going on and what type of build is being requested. The same cmake process is used locally and in the cloud (circleci, appveyor and travis), this allows the building and packaging locally the same as will be done remotely. It also allows the full process to be tested and the packages needed by plugin manager to be generated and used.

If you do a git clone or fork of the plugin then it should build. If you are stripping out the git information you are going to be on your own. Currently OCPN does not build on OpenSUSE in the cloud so unless you are adding that capability the plugins will not have it. Fedora is s build environment, so any distribution which is based on Fedora should be able to use both OCPN and the plugins and if frontend2 is being used then rpms are generated and loaded into cloudsmith so legacy installation is possible.

Do the rpms that are currently generated not work?

**Comment by dominig** [comment]
John,
OpenSUSE as well Fedora build tools get the source code from Git but they provide their own controlled build system.
In that model OpenCPN is provided and maintain by the distro under the responsibility of the distro which offers a coherent set of code between the distro and the application.
The plugins in that mode are installed under root in section of the file system which is not accessible by regular user and the application are recompiled, updated and re-dispatched each time that a change in any middleware risk to create some incompatibilities.
Extracting the git tag to introduce it into the Distro build system is not something avaialble (when you look at the version of Firefox or KDE, you get a version number not a git tag which is abscon to the user.
For debug, the git tag is included in rpm detailed name by the publication process but is not visible by cmake.

What I propose is just to support LEGACY build mode that allows me to provide OpenCPN as part of the official package of the distro, what allows to keep it updated under control of the distro with a reduced volume of data (as partial rpm are created) and a decent security (as all packages are signed). I am open of the best way to offer it but I would rather do it without having to make a fork.

**Comment by rgleason** [comment]
Dominig, using the packages from Cloudsmith and the "old way" unmanaged plugin is a perfectly reasonable approach, but in that case, why would any change be needed in Frontend2, the packages are all provided.

I hope you are not asking Jon to turn Frontend22 into a more complicated pretzel of code to support opensuse. Perhaps there are some small specific things that are needed? If you could advise in more detail perhaps that will help Jon field this request and perhaps meet your goal.

**Comment by jongough** [comment]
Is there some identifier that is supplied as an 'environment variable' like travis, circleci and appveyor do so that the build process knows where it is running? Then it may be possible to bypass any git stuff, although I am still not sure how you get the full directory structure and make sure it is up to date if git is not being used to get the structure and hence have a '.git' directory.

**Comment by jongough** [comment]
Looking at the OBS system for OpenSUSE it appears to be quite out of date with OCPN. OCPN is now 5.2.x . 

It would appear that a spec file is needed to build the plugin, not quite sure where or how this is maintained as it appears to be separate from the source repository. There needs to be a way to extract the whole source tree of a particular branch into a zip file for it to be used. This is can be done, but it is yet another process that is required. The whole thrust of what has been done in the last year is to automate as much as possible such that pull requests, tags and releases automatically cause builds to occur, these builds deploy to known locations. There is still a manual process for updating the Plugin Manger xml file, but it is trivial. OCPN can then find the new version of plugins and install them in user writable areas. For the OBS to be used it would have to be automated and included in the current process, i.e. on each git update (pull request, tag, etc.) start a process somewhere that does the extract, causes the OBS system to import the extract, then build it, all without the user having to do anything.  If you know how to do this then please create a clone of testplugin_pi, create a new branch, make the updates, test it then submit a pull request and we will see how it goes.

**Comment by dominig** [comment]
John, the official OpenSUSE repo is at 5.0 because many plugin do not build with 5.2. The development project which is on my home page is already at 5.2 but many plugins are not available.
**Source code update**
The OBS upload the code from git when I trigger and update. I stick to the stable build of OpenCPN and plugins. I can force a specific git branch or tag in OBS (via the _service file)
**Spec Files maintenance**
I have been the official maintainer of the spec file for the last 10 years and kept an OpenCPN with plugin available for the OpenSUSE community during all these years. I work with 2 OBS projects, a sandbox (in my home zone https://build.opensuse.org/project/show/home:dominig:opencpn-sandbox) for checking latest code drop and a release one (in the official zone https://build.opensuse.org/project/show/Application:Geo:OpenCPN) that is published only after full test on the official repo. I only provide a development version for the running OpenSUSE distro (Tumbleweed) as such unstable environment is not suitable at sea. Official release oare only done for the stable branch of the distro (Leap).
The "same" spec files are reused by the Linux Marine grade distribution that will go public this week. The building process is Koji and has the same requirement than an OBS.
 https://redpesk.bzh/
**Spec file on GitHub**
Spec files could be maintained in OpenCPN git hub repo if desired but as they are specific to OpenSUE and Fedora and not easy to maintain by main code developers. I maintain them in their respective building tools which offer source control and are public.
an example of how to do it can be found in one of the other project where I was architect in my paid job (Intel).
https://build.opensuse.org/package/view_file/Cloud:StarlingX:2.0/cgts-client/_service?expand=1
**XML files and Pluging download**
Due to the building process, I do not need to worry about possible dependency incompatibility between third party library, Global consistency is enforced between the building and the controlled update process of the distro.
This process has been running and maintained for 10 years and has proven to be safe with over 30000 packages for OpenSUSE alone.
As the plugins are maintain by the distro, there is not need to request OpenCPN to download the plugin update, that is done by the regular Distro updates. I have always installed all the running plugin by default, letting the user to activate them or not as they wish.
**Auto build on each GitHub commit**
I could activate a hook in GitHub to automatically trigger an OBS build on each source code commit but that is honestly present very little advantage, as I do not publish any code in the official repo without at least a health check. Building silently present little advantages. Would each plugin be provided with a test harness that would be different.
**What I need is simple**
In short the only thing that I request is a building option to build in rpmbuild which does not run the build in a git tree.
This is what I have demonstrated in my fork of climatology_pi as proof of concept by adding a LEGACY_BUILD option in cmake.
There no need to change anything else. I may create a patch in OpenCPN to deactivate the plugin download and replace it by a message clarifying that it's done by the standard distro update (zypper for OpenCPN or dnf for Fedora).
I would rather have the option available for anyone rather than to create my private forks, this is why I am trying to sync with all of you.

**Comment by dominig** [comment]
> Dominig, using the packages from Cloudsmith and the legacy unmanaged plugin is a perfectly reasonable approach, but in that case, why would any change be needed in Frontend2, the packages are all provided.
> 
> https://cloudsmith.io/~opencpn/repos/celestial-navigation-alpha/packages/
> 
> I hope you are not asking Jon to turn Frontend22 into a more complicated pretzel of code to support opensuse. Perhaps there are some small specific things that are needed? If you could advise in more detail perhaps that will help Jon field this request and perhaps meet your goal.
**Resp**
No I do not ask any change at that level as  do not use it at all.

**Comment by rgleason** [comment]
Dominiq,
Your Simple Request....."in my fork of climatology_pi as proof of concept by adding a LEGACY_BUILD option in cmake."

See this which has confilicts and is too many changes, please.  Can you use my master branch fork of Sean's plugin to make a PR with just the Legacy_Build option?   Thanks

**Comment by jongough** [comment]
The name "BUILD_LEGACY_MODE" is not a good choice. The current frontend2 build process does actually build the legacy installables, i.e. deb, exe, etc.. What you want is an OpenSUSE build type such that it builds using the OBS and no git. The variable name should reflect the OpenSUSE environment as it is specific to that environment only.

I was suggesting that there may/should be an environment variable specific to the OBS system that could be used to change the build process. Currently Travis, Appveyor and Circleci are handled as they have an environment variable which Cmake can check to see what sort of build is required. Flatpak is another build environment which needs identifying and it has its own path through the cmake files. These checks are then placed in the frontend2 cmake code that the general developer does not need to change, i.e. they can copy newer versions of frontend2 over the top of the current files. The whole thrust of frontend2 is to make all user modifiable parts be in CMakeLists.txt with all the rest being support functionality. The user can add extra cmake capabilities that may be required by plugins easily. To this end the OpenSUSE build part should be hidden and not contained in CMakeLists.txt or we start to go back to complex cmake code the developer has to maintain.

If you can provide the name of the environment variable that uniquely identifies the OBS build process I can add that to frontend2.

**Comment by rgleason** [comment]
Dominig, one other thing you should be aware of, Jon's plugin  TestPlugin_pi which is the Frontend2 "Parent" to all of Sean's plugins.  Changes made there are generally propagated to Sean's plugins. Perhaps your PR should be using that plugin as a start?

**Comment by dominig** [comment]
Thanks for the advise. As I do not provide that test plugin, I have not look at it yet but I will soon.

**Comment by rgleason** [comment]
Dominig Can this be closed?
